go-controller/pkg/node/node.go

```diff
@@ -286,54 +292,47 @@ func (n *OvnNode) WatchEndpoints() {
 		UpdateFunc: func(old, new interface{}) {
 			epNew := new.(*kapi.Endpoints)
 			epOld := old.(*kapi.Endpoints)
+			// return if there are no endpoint changes
 			if apiequality.Semantic.DeepEqual(epNew.Subsets, epOld.Subsets) {
 				return
 			}
-			newEpAddressMap := buildEndpointAddressMap(epNew.Subsets)
-			for item := range buildEndpointAddressMap(epOld.Subsets) {
-				if _, ok := newEpAddressMap[item]; !ok {
-					err := deleteConntrack(item.ip, item.port, item.protocol)
-					if err != nil {
-						klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
-					}
-				}
+			// get the service that belongs to the endpoint
+			svc, err := n.watchFactory.GetService(epNew.Namespace, epNew.Name)
+			if err != nil {
+				klog.Errorf("Error getting service %s/%s after its endpoints changed: %v", epNew.Name, epNew.Namespace, err)
+				return
 			}
+			deleteServiceConntrackEntries(svc)
 		},
 		DeleteFunc: func(obj interface{}) {
 			ep := obj.(*kapi.Endpoints)
-			for item := range buildEndpointAddressMap(ep.Subsets) {
-				err := deleteConntrack(item.ip, item.port, item.protocol)
-				if err != nil {
-					klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
-				}
-
+			// get the service that belongs to the endpoint
+			svc, err := n.watchFactory.GetService(ep.Namespace, ep.Name)
+			if err != nil {
+				klog.Errorf("Error getting service %s/%s after its endpoints changed: %v", ep.Name, ep.Namespace, err)
+				return
 			}
+			deleteServiceConntrackEntries(svc)
 		},
 	}, nil)
 }
 
-type epAddressItem struct {
-	ip       string
-	port     int32
-	protocol kapi.Protocol
-}
+func deleteServiceConntrackEntries(svc *kapi.Service) {
+	// return if the service doesn't have ClusterIPs
+	if !util.IsClusterIPSet(svc) {
+		return
+	}
 
-//buildEndpointAddressMap builds a map of all UDP and SCTP ports in the endpoint subset along with that port's IP address
-func buildEndpointAddressMap(epSubsets []kapi.EndpointSubset) map[epAddressItem]struct{} {
-	epMap := make(map[epAddressItem]struct{})
-	for _, subset := range epSubsets {
-		for _, address := range subset.Addresses {
-			for _, port := range subset.Ports {
-				if port.Protocol == kapi.ProtocolUDP || port.Protocol == kapi.ProtocolSCTP {
-					epMap[epAddressItem{
-						ip:       address.IP,
-						port:     port.Port,
-						protocol: port.Protocol,
-					}] = struct{}{}
-				}
+	for _, ip := range util.GetClusterIPs(svc) {
+		for _, port := range svc.Spec.Ports {
+			// only connectionless protocols can leave stale conntrack connections
+			if port.Protocol != kapi.ProtocolUDP && port.Protocol != kapi.ProtocolSCTP {
+				continue
+			}
+			err := deleteConntrack(ip, port.Port, port.Protocol)
```

Comment:
I am going to review this PR later, but since I happened to take a look at this commit:

previously we deleted conntrack entries based on the endpoint's `port`, this is equivalent to a service `targetPort`, right? So shouldn't you do `err := deleteConntrack(ip, port.TargetPort, port.Protocol)` here for the functionality to remain the same?